tables/extract.py

```diff
@@ -0,0 +1,131 @@
+from collections import defaultdict
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+
+from tables.table import Table, Tag
+
+
+class TableExtractor:
+    """
+    Extracts tables from a workbook based on specified ID tags.
+
+    This class provides methods to read data from CSV or Excel files and extract tables
+    based on unique ID tags. It identifies the positions of start and end tags in the worksheet,
+    pairs them correctly, and extracts the table located between them.
+
+    Attributes:
+        START_TAG (str): The start tag format for identifying tables.
+        END_TAG (str): The end tag format for identifying tables.
+        workbook (dict[str, pd.DataFrame]): A dictionary containing worksheet names as keys
+            and corresponding pandas DataFrames as values.
+
+    Methods:
+        from_csv(cls, path: Path, worksheet_name: str) -> "TableExtractor":
+            Creates a TableExtractor instance from a CSV file.
+        from_excel(cls, path: Path) -> "TableExtractor":
+            Creates a TableExtractor instance from an Excel file.
+        extract(self, worksheet_name: str, id_tag: str) -> list[Table]:
+            Extracts tables specified by the given ID tag.
+
+    Raises:
+        ValueError: If not all ID tags have matching start and end tags.
+
+    Example usage:
+    >>> extractor = TableExtractor.from_excel(path=...)
+    >>> tables = extractor.extract("Sheet1", "Table1")
+    """
+
+    START_TAG = "{id}S"
+    END_TAG = "{id}E"
+    workbook: dict[str, pd.DataFrame]
+
+    def __init__(self, workbook: dict[str, pd.DataFrame]) -> None:
+        self.workbook = workbook
+
+    @classmethod
+    def from_csv(cls, path: Path, worksheet_name: str) -> "TableExtractor":
+        return cls(workbook={worksheet_name: pd.read_csv(path, header=None, index_col=None)})
+
+    @classmethod
+    def from_excel(cls, path: Path) -> "TableExtractor":
+        return cls(workbook=pd.read_excel(path, index_col=None, header=None, sheet_name=None))
+
+    def extract(self, worksheet_name: str, id_tag: str) -> list[Table]:
+        """Extracts table instances specified by ID.
+
+        Finds IDs in the worksheet that are positioned above the top left and below the bottom right cells of the tables
+        and correctly pairs them together. Using those positions, extracts the table located between them.
+
+        :param worksheet_name: worksheet to extract tables from
+        :param id_tag: ID of table to locate and extract
+        :return: a set of Table objects
+        """
+        worksheet = self.workbook[worksheet_name]
+        end_tags, start_tags = self._get_tags(id_tag, worksheet)
+        paired_tags = self._pair_tags(start_tags, end_tags, file_width=len(worksheet.columns))
+        dfs = self._extract_dfs(worksheet, paired_tags)
+        tables = [Table(df=df, start_tag=start_tag, id_tag=id_tag) for df, (start_tag, _) in zip(dfs, paired_tags)]
```

Comment:
Fair enough, can we remove it now though please as not functionally required?

Reply:
I would be in favour of keeping it in because it's useful even if it's not functionally required. I'd argue it adds negligible complexity and is still within the theme of the `Table` class (supplementing a dataframe with context from the original source)